vispy/visuals/markers.py

```diff
@@ -27,95 +28,169 @@
 varying vec4 v_fg_color;
 varying vec4 v_bg_color;
 varying float v_edgewidth;
-varying float v_antialias;
+varying float v_depth_middle;
+varying float v_alias_ratio;
+
+float big_float = 1e10; // prevents numerical imprecision
 
 void main (void) {
-    $v_size = a_size * u_px_scale * u_scale;
-    v_edgewidth = a_edgewidth * float(u_px_scale);
-    v_antialias = u_antialias;
     v_fg_color  = a_fg_color;
     v_bg_color  = a_bg_color;
-    gl_Position = $transform(vec4(a_position,1.0));
-    float edgewidth = max(v_edgewidth, 1.0);
-    gl_PointSize = ($v_size) + 4.*(edgewidth + 1.5*v_antialias);
+
+    vec4 pos = vec4(a_position, 1);
+    vec4 fb_pos = $visual_to_framebuffer(pos);
+    gl_Position = $framebuffer_to_render(fb_pos);
+
+    // NOTE: gl_stuff uses framebuffer coords!
+
+    if (u_scaling == true) {
+        // calculate point size from visual to framebuffer coords to determine size
+        vec4 x = $framebuffer_to_visual(fb_pos + vec4(big_float, 0, 0, 0));
+        x = (x - pos);
+        vec4 size_vec = $visual_to_framebuffer(pos + normalize(x) * a_size);
+        $v_size = size_vec.x / size_vec.w - fb_pos.x / fb_pos.w;
+        v_edgewidth = ($v_size / a_size) * a_edgewidth;
+    }
+    else {
+        $v_size = a_size * u_px_scale;
+        v_edgewidth = a_edgewidth * u_px_scale;
+    }
+
+    // gl_PointSize is the diameter
+    gl_PointSize = $v_size + 4. * (v_edgewidth + 1.5 * u_antialias);
+
+    if (u_sphere == true) {
```

Comment:
Could this be named and treated in a more generic way? Like maybe this isn't just `sphere` but is more of a "3D markers" effect? So `u_3d` or something? Overall you are finding some Z distance right?

Reply:
I did think about this, but then focused on getting this to work first :P I think it's actually doable and not too much work... whtether it will look good, it's another question :P The result would be that any marker would look like a "carved out" sphere that always faces the screen. Anything beyond a sphere radius would look terrible, though, because it's flattened out to the same depth and lighting value. Maybe that can be solved as well, but it's definitely harder.

Reply:
I did it! It was trivial, in the end :) It does look pretty decent on most markers (square ones are a bit weird on the edges, but oh well).


https://user-images.githubusercontent.com/23482191/131989560-dbf11cc6-1528-4c91-80f9-778dbac5c84b.mp4